app/frontend/wallet/helpers/generateDeriveXpubFn.js

```diff
@@ -0,0 +1,38 @@
+const indexIsHardened = require('./indexIsHardened')
+const {derivePublic} = require('cardano-crypto.js')
+
+function generateDeriveXpubFn(deriveXpubHardenedFn, derivationScheme) {
+  const derivedXpubs = {}
+
+  async function deriveXpub(absDerivationPath) {
+    const memoKey = JSON.stringify(absDerivationPath)
+
+    if (!derivedXpubs[memoKey]) {
+      const deriveHardened =
+        absDerivationPath.length === 0 ||
+        indexIsHardened(absDerivationPath[absDerivationPath.length - 1])
+
+      /*
+      * TODO - reset cache if the promise fails, for now it does not matter
+      * since a failure (e.g. rejection by user) there leads to
+      * the creation of a fresh wallet and crypto provider instance
+      */
+      derivedXpubs[memoKey] = deriveHardened
+        ? deriveXpubHardenedFn(absDerivationPath)
+        : deriveXpubNonhardenedFn(absDerivationPath)
+    }
+
+    return await derivedXpubs[memoKey]
+  }
+
+  async function deriveXpubNonhardenedFn(derivationPath) {
+    const lastIndex = derivationPath[derivationPath.length - 1]
+    const parentXpub = await deriveXpub(derivationPath.slice(0, derivationPath.length - 1))
+
+    return derivePublic(parentXpub, lastIndex, derivationScheme.number)
```

Comment:
Hmm, I don't like `derivePublic` name as it works only one step. What about `deriveChildPubkey` or something like that?

Reply:
Hmm, more general question -- why does this function accept `derivationScheme.number` ? I thought that HDnode derivation is same between schemes, it is just the seed that is different

Reply:
https://github.com/input-output-hk/cardano-crypto/blob/master/cbits/encrypted_sign.c#L362 - the derivation step also depends on the scheme